contributors/design-proposals/storage/container-storage-interface.md

```diff
@@ -444,34 +444,122 @@ Alternatively, deployment could be simplified by having all components (includin
 This part of proposal is not going to be implemented in alpha release.
 
 #### End user credentials
-CSI allows specifying *end user credentials* in all operations. Kubernetes does not have facility to configure a Secret per *user*, we usually track objects per *namespace*. Therefore we decided to postpone implementation of these credentials and wait until CSI is clarified.
+CSI allows specifying credentials in CreateVolume/DeleteVolume, ControllerPublishVolume/ControllerUnpublishVolume, and NodePublishVolume/NodeUnpublishVolume operations.
+
+Kubernetes will enable cluster admins and application developers to specify these credentials by referencing Kubernetes secret object(s). Kubernetes (either the core components or helper containers) will fetch the secret(s) and pass them to the CSI volume plugin.
+
+#### CreateVolume/DeleteVolume Credentials
+
+```yaml
+kind: StorageClass
+apiVersion: storage.k8s.io/v1
+metadata:
+  name: fast-storage
+provisioner: com.example.team.csi-driver
+parameters:
+  type: pd-ssd
+  csiProvisionerSecretName: mysecret
+  csiProvisionerSecretNamespace: mynamespaace
+```
+
+The CSI external-provisioner will reserve the parameter keys `csiProvisionerSecretName` and `csiProvisionerSecretNamespace`. If specified, the CSI Provisioner will fetch the secret `csiProvisionerSecretName` in the Kubernetes namespace `csiProvisionerSecretNamespace` and pass it to:
+1. The CSI `CreateVolumeRequest` in the `controller_create_credentials` field.
+2. The CSI `DeleteVolumeRequest` in the `controller_delete_credentials` field.
```

Comment:
I don't see the `controller_create_credentials` field in these structs... what is the data type?

Reply:
@liggitt Both mentioned fields are at the following links:
https://github.com/container-storage-interface/spec/blob/master/lib/go/csi/csi.pb.go#L520
https://github.com/container-storage-interface/spec/blob/master/lib/go/csi/csi.pb.go#L886

Reply:
ah, thanks, I was looking at the CSI types vendored in kube

since Secret objects can contain binary data, how would non-utf-8 data be coerced into `map[string]string`? if that was encountered, what would the provisioner do? corrupt the data, transform into base64 strings, fail the provision request, or something else?

Reply:
@liggitt Would unconditionally converting secret's data (string or binary) to base64 before assigning to controller_*_credentials and then converting it back not be an acceptable solution?

Reply:
Per the spec, CSI will only except string data. But a plugin can dictate in its documentation that a specific secret contain binary data and specify what binary-to-text encoding to use (base64, quoted-printable, etc.). It is the responsibility of the entity (cluster admin, user, etc.) that creates the secret, in that case, to ensure its content is what the plugin expects and is encoded in the format the plugin expects.

Reply:
I will clarify this in the doc.

Reply:
> Would unconditionally converting secret's data (string or binary) to base64 before assigning to controller_*_credentials and then converting it back by a plugin not be an acceptable solution?

no... I'd expect transparent plumbing of Secret data in. that likely means limiting the accepted secrets to ones containing data that fits in the string type without corruption.

> a plugin can dictate in its documentation that a specific secret contain binary data and specify what binary-to-text encoding to use (base64, quoted-printable, etc.). It is the responsibility of the entity (cluster admin, user, etc.) that creates the secret, in that case, to ensure its content is what the plugin expects and is encoded in the format the plugin expects.

I agree, just wanted to spec out the behavior if the provisioner encountered a secret containing binary data